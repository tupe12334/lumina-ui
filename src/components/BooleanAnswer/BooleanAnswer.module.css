.container {
  display: flex;
  align-items: center;
  gap: 12px;
}

.container.disabled {
  opacity: 0.75;
}

.button {
  min-width: 100px;
  border-radius: 12px;
  border: 2px solid;
  font-weight: 500;
  cursor: pointer;
  transition: all 0.2s ease;
  background-color: #ffffff;
  border-color: #d1d5db;
  color: #374151;
}

.button:disabled {
  cursor: not-allowed;
  opacity: 0.5;
}

.button.selected {
  color: #ffffff;
}

.button.selected.true {
  background-color: #22c55e;
  border-color: #22c55e;
}

.button.selected.false {
  background-color: #ef4444;
  border-color: #ef4444;
}

.button:not(.selected):not(:disabled):hover {
  border-color: #9ca3af;
}

.button.small {
  padding: 8px 16px;
  font-size: 14px;
}

.button.medium {
  padding: 10px 20px;
  font-size: 16px;
}

.button.large {
  padding: 12px 24px;
  font-size: 18px;
}