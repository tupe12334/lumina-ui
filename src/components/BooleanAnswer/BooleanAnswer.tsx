import React, { useState } from 'react'

export interface BooleanAnswerProps {
  /**
   * Initial value for the boolean answer
   */
  initialValue?: boolean | null
  /**
   * Callback when user selects a boolean value
   */
  onSelect?: (value: boolean) => void
  /**
   * Label for the true/positive option
   */
  trueLabel?: string
  /**
   * Label for the false/negative option
   */
  falseLabel?: string
  /**
   * Whether the component is disabled
   */
  disabled?: boolean
  /**
   * Additional CSS classes
   */
  className?: string
  /**
   * Size of the buttons
   */
  size?: 'sm' | 'md' | 'lg'
}

/**
 * Boolean answer component for Yes/No and True/False questions.
 * Automatically detects question type and provides appropriate button labels.
 */
export const BooleanAnswer = React.forwardRef<HTMLDivElement, BooleanAnswerProps>(
  ({
    initialValue,
    onSelect,
    trueLabel = 'True',
    falseLabel = 'False',
    disabled = false,
    className,
    size = 'md',
    ...props
  }, ref) => {
    if (
      initialValue !== undefined &&
      initialValue !== null &&
      typeof initialValue !== 'boolean'
    ) {
      throw new Error(
        'BooleanAnswer initialValue must be boolean, null, or undefined'
      )
    }

    const [selectedValue, setSelectedValue] = useState<boolean | null>(
      initialValue !== null && initialValue !== undefined ? initialValue : null
    )

    const handleSelect = (value: boolean) => {
      if (disabled) return

      setSelectedValue(value)
      onSelect && onSelect(value)
    }


    const containerStyle: React.CSSProperties = {
      display: 'flex',
      alignItems: 'center',
      gap: '12px',
      opacity: disabled ? 0.75 : 1,
    }

    const getButtonStyle = (isSelected: boolean): React.CSSProperties => ({
      minWidth: '100px',
      padding: size === 'sm' ? '8px 16px' : size === 'lg' ? '12px 24px' : '10px 20px',
      borderRadius: '12px',
      border: '2px solid',
      fontWeight: 500,
      fontSize: size === 'sm' ? '14px' : size === 'lg' ? '18px' : '16px',
      cursor: disabled ? 'not-allowed' : 'pointer',
      transition: 'all 0.2s ease',
      backgroundColor: isSelected ? '#ef4444' : '#ffffff',
      borderColor: isSelected ? '#ef4444' : '#d1d5db',
      color: isSelected ? '#ffffff' : '#374151',
      opacity: disabled ? 0.5 : 1,
    })

    return (
      <div
        ref={ref}
        style={{...containerStyle, ...className}}
        {...props}
      >
        <button
          disabled={disabled}
          onClick={() => handleSelect(true)}
          style={getButtonStyle(selectedValue === true)}
          aria-pressed={selectedValue === true}
          aria-label={`${trueLabel} option`}
          onMouseEnter={(e) => {
            if (!disabled && selectedValue !== true) {
              e.currentTarget.style.borderColor = '#9ca3af'
            }
          }}
          onMouseLeave={(e) => {
            if (!disabled && selectedValue !== true) {
              e.currentTarget.style.borderColor = '#d1d5db'
            }
          }}
        >
          {trueLabel}
        </button>

        <button
          disabled={disabled}
          onClick={() => handleSelect(false)}
          style={getButtonStyle(selectedValue === false)}
          aria-pressed={selectedValue === false}
          aria-label={`${falseLabel} option`}
          onMouseEnter={(e) => {
            if (!disabled && selectedValue !== false) {
              e.currentTarget.style.borderColor = '#9ca3af'
            }
          }}
          onMouseLeave={(e) => {
            if (!disabled && selectedValue !== false) {
              e.currentTarget.style.borderColor = '#d1d5db'
            }
          }}
        >
          {falseLabel}
        </button>
      </div>
    )
  }
)

BooleanAnswer.displayName = 'BooleanAnswer'