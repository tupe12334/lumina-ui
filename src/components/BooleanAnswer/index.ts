export { BooleanAnswer } from './BooleanAnswer'
export type { BooleanAnswerProps } from './BooleanAnswer'