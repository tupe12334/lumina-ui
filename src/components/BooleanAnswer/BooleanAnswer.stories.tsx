import type { Meta, StoryObj } from '@storybook/react'
import { BooleanAnswer } from './BooleanAnswer'

const meta: Meta<typeof BooleanAnswer> = {
  title: 'Components/BooleanAnswer',
  component: BooleanAnswer,
  parameters: {
    layout: 'centered',
    docs: {
      description: {
        component: 'Boolean answer component for Yes/No and True/False questions. Automatically detects question type based on question text and language.',
      },
    },
  },
  tags: ['autodocs'],
  argTypes: {
    initialValue: {
      control: { type: 'select' },
      options: [true, false, null],
      description: 'Initial selected value',
    },
    onSelect: {
      action: 'selected',
      description: 'Callback when user selects a value',
    },
    trueLabel: {
      control: 'text',
      description: 'Label for the true/positive option',
    },
    falseLabel: {
      control: 'text',
      description: 'Label for the false/negative option',
    },
    disabled: {
      control: 'boolean',
      description: 'Whether the component is disabled',
    },
    size: {
      control: { type: 'select' },
      options: ['sm', 'md', 'lg'],
      description: 'Size of the buttons',
    },
  },
}

export default meta
type Story = StoryObj<typeof meta>

// Basic True/False question
export const TrueFalse: Story = {
  args: {
    trueLabel: 'True',
    falseLabel: 'False',
    size: 'md',
  },
}

// Yes/No question
export const YesNo: Story = {
  args: {
    trueLabel: 'Yes',
    falseLabel: 'No',
    size: 'md',
  },
}

// Pre-selected True
export const PreSelectedTrue: Story = {
  args: {
    trueLabel: 'True',
    falseLabel: 'False',
    initialValue: true,
    size: 'md',
  },
}

// Pre-selected False
export const PreSelectedFalse: Story = {
  args: {
    trueLabel: 'True',
    falseLabel: 'False',
    initialValue: false,
    size: 'md',
  },
}

// Small size
export const Small: Story = {
  args: {
    trueLabel: 'Yes',
    falseLabel: 'No',
    size: 'sm',
  },
}

// Large size
export const Large: Story = {
  args: {
    trueLabel: 'Yes',
    falseLabel: 'No',
    size: 'lg',
  },
}

// Hebrew Labels
export const HebrewLabels: Story = {
  args: {
    trueLabel: 'כן',
    falseLabel: 'לא',
    size: 'md',
  },
  parameters: {
    docs: {
      description: {
        story: 'Hebrew labels example',
      },
    },
  },
}

// Disabled state
export const Disabled: Story = {
  args: {
    trueLabel: 'True',
    falseLabel: 'False',
    initialValue: true,
    disabled: true,
    size: 'md',
  },
}

// Direct Labels (Recommended)
export const DirectLabels: Story = {
  args: {
    trueLabel: 'Agree',
    falseLabel: 'Disagree',
    size: 'md',
  },
  parameters: {
    docs: {
      description: {
        story: 'Using direct labels instead of automatic detection - the recommended approach',
      },
    },
  },
}

// Custom Labels
export const CustomLabels: Story = {
  args: {
    trueLabel: 'Accept',
    falseLabel: 'Decline',
    size: 'md',
  },
}

// Interactive example
export const Interactive: Story = {
  args: {
    trueLabel: 'Yes',
    falseLabel: 'No',
    size: 'md',
    onSelect: (value: boolean) => {
      console.log('Selected:', value)
      // In a real app, you might update state, send analytics, etc.
    },
  },
  parameters: {
    docs: {
      description: {
        story: 'Click the buttons to see the callback in action (check browser console)',
      },
    },
  },
}

// Showcase different label types
export const LabelTypes: Story = {
  render: () => (
    <div className="space-y-8 p-6">
      <div>
        <h3 className="text-lg font-semibold mb-4 text-center">True/False</h3>
        <BooleanAnswer
          trueLabel="True"
          falseLabel="False"
          size="md"
          onSelect={(value) => console.log('True/False:', value)}
        />
      </div>

      <div>
        <h3 className="text-lg font-semibold mb-4 text-center">Yes/No</h3>
        <BooleanAnswer
          trueLabel="Yes"
          falseLabel="No"
          size="md"
          onSelect={(value) => console.log('Yes/No:', value)}
        />
      </div>

      <div>
        <h3 className="text-lg font-semibold mb-4 text-center">Agree/Disagree</h3>
        <BooleanAnswer
          trueLabel="Agree"
          falseLabel="Disagree"
          size="md"
          onSelect={(value) => console.log('Agree/Disagree:', value)}
        />
      </div>
    </div>
  ),
  parameters: {
    docs: {
      description: {
        story: 'Demonstration of different label types using direct props',
      },
    },
  },
}

// All sizes comparison
export const AllSizes: Story = {
  render: () => (
    <div className="space-y-6 p-6">
      <div>
        <h3 className="text-sm font-medium mb-3 text-center">Small</h3>
        <BooleanAnswer
          trueLabel="Yes"
          falseLabel="No"
          size="sm"
        />
      </div>

      <div>
        <h3 className="text-base font-medium mb-3 text-center">Medium (Default)</h3>
        <BooleanAnswer
          trueLabel="Yes"
          falseLabel="No"
          size="md"
        />
      </div>

      <div>
        <h3 className="text-lg font-medium mb-3 text-center">Large</h3>
        <BooleanAnswer
          trueLabel="Yes"
          falseLabel="No"
          size="lg"
        />
      </div>
    </div>
  ),
  parameters: {
    docs: {
      description: {
        story: 'Comparison of all available sizes',
      },
    },
  },
}