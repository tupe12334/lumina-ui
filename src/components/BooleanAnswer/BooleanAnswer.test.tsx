import { render } from '@testing-library/react'
import { test } from 'vitest'
import { BooleanAnswer } from './BooleanAnswer'

test('BooleanAnswer renders correctly with default props', () => {
  const { container } = render(<BooleanAnswer />)
  expect(container.firstChild).toMatchSnapshot()
})

test('BooleanAnswer renders correctly with custom labels', () => {
  const { container } = render(
    <BooleanAnswer trueLabel="Yes" falseLabel="No" />
  )
  expect(container.firstChild).toMatchSnapshot()
})

test('BooleanAnswer renders correctly with initial value true', () => {
  const { container } = render(<BooleanAnswer initialValue={true} />)
  expect(container.firstChild).toMatchSnapshot()
})

test('BooleanAnswer renders correctly with initial value false', () => {
  const { container } = render(<BooleanAnswer initialValue={false} />)
  expect(container.firstChild).toMatchSnapshot()
})

test('BooleanAnswer renders correctly with initial value null', () => {
  const { container } = render(<BooleanAnswer initialValue={null} />)
  expect(container.firstChild).toMatchSnapshot()
})

test('BooleanAnswer renders correctly when disabled', () => {
  const { container } = render(<BooleanAnswer disabled />)
  expect(container.firstChild).toMatchSnapshot()
})

test('BooleanAnswer renders correctly with small size', () => {
  const { container } = render(<BooleanAnswer size="sm" />)
  expect(container.firstChild).toMatchSnapshot()
})

test('BooleanAnswer renders correctly with medium size', () => {
  const { container } = render(<BooleanAnswer size="md" />)
  expect(container.firstChild).toMatchSnapshot()
})

test('BooleanAnswer renders correctly with large size', () => {
  const { container } = render(<BooleanAnswer size="lg" />)
  expect(container.firstChild).toMatchSnapshot()
})

test('BooleanAnswer renders correctly with custom className', () => {
  const { container } = render(<BooleanAnswer className="custom-class" />)
  expect(container.firstChild).toMatchSnapshot()
})

test('BooleanAnswer renders correctly with Hebrew labels', () => {
  const { container } = render(
    <BooleanAnswer trueLabel="כן" falseLabel="לא" />
  )
  expect(container.firstChild).toMatchSnapshot()
})

test('BooleanAnswer renders correctly with Agree/Disagree labels', () => {
  const { container } = render(
    <BooleanAnswer trueLabel="Agree" falseLabel="Disagree" />
  )
  expect(container.firstChild).toMatchSnapshot()
})

test('BooleanAnswer renders correctly with all props combined', () => {
  const { container } = render(
    <BooleanAnswer
      trueLabel="Accept"
      falseLabel="Decline"
      initialValue={true}
      disabled
      size="lg"
      className="custom-class"
      data-testid="complex-boolean-answer"
    />
  )
  expect(container.firstChild).toMatchSnapshot()
})

test('BooleanAnswer renders correctly with onSelect callback', () => {
  const handleSelect = () => {}
  const { container } = render(
    <BooleanAnswer onSelect={handleSelect} />
  )
  expect(container.firstChild).toMatchSnapshot()
})