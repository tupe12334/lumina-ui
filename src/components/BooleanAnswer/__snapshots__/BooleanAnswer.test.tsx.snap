// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`BooleanAnswer renders correctly when disabled 1`] = `
<div
  class="_container_83693f _disabled_83693f"
>
  <button
    aria-label="True option"
    aria-pressed="false"
    class="_button_83693f _medium_83693f"
    disabled=""
  >
    True
  </button>
  <button
    aria-label="False option"
    aria-pressed="false"
    class="_button_83693f _medium_83693f"
    disabled=""
  >
    False
  </button>
</div>
`;

exports[`BooleanAnswer renders correctly with Agree/Disagree labels 1`] = `
<div
  class="_container_83693f"
>
  <button
    aria-label="Agree option"
    aria-pressed="false"
    class="_button_83693f _medium_83693f"
  >
    Agree
  </button>
  <button
    aria-label="Disagree option"
    aria-pressed="false"
    class="_button_83693f _medium_83693f"
  >
    Disagree
  </button>
</div>
`;

exports[`BooleanAnswer renders correctly with Hebrew labels 1`] = `
<div
  class="_container_83693f"
>
  <button
    aria-label="כן option"
    aria-pressed="false"
    class="_button_83693f _medium_83693f"
  >
    כן
  </button>
  <button
    aria-label="לא option"
    aria-pressed="false"
    class="_button_83693f _medium_83693f"
  >
    לא
  </button>
</div>
`;

exports[`BooleanAnswer renders correctly with all props combined 1`] = `
<div
  class="_container_83693f _disabled_83693f custom-class"
  data-testid="complex-boolean-answer"
>
  <button
    aria-label="Accept option"
    aria-pressed="true"
    class="_button_83693f _large_83693f _selected_83693f _true_83693f"
    disabled=""
  >
    Accept
  </button>
  <button
    aria-label="Decline option"
    aria-pressed="false"
    class="_button_83693f _large_83693f"
    disabled=""
  >
    Decline
  </button>
</div>
`;

exports[`BooleanAnswer renders correctly with custom className 1`] = `
<div
  class="_container_83693f custom-class"
>
  <button
    aria-label="True option"
    aria-pressed="false"
    class="_button_83693f _medium_83693f"
  >
    True
  </button>
  <button
    aria-label="False option"
    aria-pressed="false"
    class="_button_83693f _medium_83693f"
  >
    False
  </button>
</div>
`;

exports[`BooleanAnswer renders correctly with custom labels 1`] = `
<div
  class="_container_83693f"
>
  <button
    aria-label="Yes option"
    aria-pressed="false"
    class="_button_83693f _medium_83693f"
  >
    Yes
  </button>
  <button
    aria-label="No option"
    aria-pressed="false"
    class="_button_83693f _medium_83693f"
  >
    No
  </button>
</div>
`;

exports[`BooleanAnswer renders correctly with default props 1`] = `
<div
  class="_container_83693f"
>
  <button
    aria-label="True option"
    aria-pressed="false"
    class="_button_83693f _medium_83693f"
  >
    True
  </button>
  <button
    aria-label="False option"
    aria-pressed="false"
    class="_button_83693f _medium_83693f"
  >
    False
  </button>
</div>
`;

exports[`BooleanAnswer renders correctly with initial value false 1`] = `
<div
  class="_container_83693f"
>
  <button
    aria-label="True option"
    aria-pressed="false"
    class="_button_83693f _medium_83693f"
  >
    True
  </button>
  <button
    aria-label="False option"
    aria-pressed="true"
    class="_button_83693f _medium_83693f _selected_83693f _false_83693f"
  >
    False
  </button>
</div>
`;

exports[`BooleanAnswer renders correctly with initial value null 1`] = `
<div
  class="_container_83693f"
>
  <button
    aria-label="True option"
    aria-pressed="false"
    class="_button_83693f _medium_83693f"
  >
    True
  </button>
  <button
    aria-label="False option"
    aria-pressed="false"
    class="_button_83693f _medium_83693f"
  >
    False
  </button>
</div>
`;

exports[`BooleanAnswer renders correctly with initial value true 1`] = `
<div
  class="_container_83693f"
>
  <button
    aria-label="True option"
    aria-pressed="true"
    class="_button_83693f _medium_83693f _selected_83693f _true_83693f"
  >
    True
  </button>
  <button
    aria-label="False option"
    aria-pressed="false"
    class="_button_83693f _medium_83693f"
  >
    False
  </button>
</div>
`;

exports[`BooleanAnswer renders correctly with large size 1`] = `
<div
  class="_container_83693f"
>
  <button
    aria-label="True option"
    aria-pressed="false"
    class="_button_83693f _large_83693f"
  >
    True
  </button>
  <button
    aria-label="False option"
    aria-pressed="false"
    class="_button_83693f _large_83693f"
  >
    False
  </button>
</div>
`;

exports[`BooleanAnswer renders correctly with medium size 1`] = `
<div
  class="_container_83693f"
>
  <button
    aria-label="True option"
    aria-pressed="false"
    class="_button_83693f _medium_83693f"
  >
    True
  </button>
  <button
    aria-label="False option"
    aria-pressed="false"
    class="_button_83693f _medium_83693f"
  >
    False
  </button>
</div>
`;

exports[`BooleanAnswer renders correctly with onSelect callback 1`] = `
<div
  class="_container_83693f"
>
  <button
    aria-label="True option"
    aria-pressed="false"
    class="_button_83693f _medium_83693f"
  >
    True
  </button>
  <button
    aria-label="False option"
    aria-pressed="false"
    class="_button_83693f _medium_83693f"
  >
    False
  </button>
</div>
`;

exports[`BooleanAnswer renders correctly with small size 1`] = `
<div
  class="_container_83693f"
>
  <button
    aria-label="True option"
    aria-pressed="false"
    class="_button_83693f _small_83693f"
  >
    True
  </button>
  <button
    aria-label="False option"
    aria-pressed="false"
    class="_button_83693f _small_83693f"
  >
    False
  </button>
</div>
`;
